plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -130,6 +230,141 @@ class bp_connection_manager {
              established_client_connection(new_connection) && num_established_clients() > self()->connections.get_max_client_count();
    }
 
+   // thread safe
+   // removes outdated signed bp_peers from msg
+   bool validate_gossip_bp_peers_message( gossip_bp_peers_message& msg ) {
+      if (msg.peers.empty())
+         return false;
+      // initial case, no server_addresses to validate
+      bool initial_msg = msg.peers.size() == 1 && msg.peers[0].server_address.empty();
+      if (!initial_msg) {
+         // validate structure and data of msg
+         auto valid_address = [](const std::string& addr) -> bool {
+            const auto& [host, port, type] = net_utils::split_host_port_type(addr);
+            return !host.empty() && !port.empty();
+         };
+         const gossip_bp_peers_message::bp_peer* prev = nullptr;
+         size_t num_per_producer = 0;
+         for (const auto& peer : msg.peers) {
+            if (peer.producer_name.empty())
+               return false; // invalid bp_peer data
+            if (!valid_address(peer.server_address))
+               return false; // invalid address
+            if (prev != nullptr) {
+               if (prev->producer_name == peer.producer_name) {
+                  ++num_per_producer;
+                  if (num_per_producer > max_bp_peers_per_producer)
+                     return false; // more than allowed per producer
+                  if (prev->server_address == peer.server_address)
+                     return false; // duplicate entries not allowed
+               } else if (prev->producer_name > peer.producer_name) {
+                  return false; // required to be sorted
+               } else {
+                  num_per_producer = 0;
+               }
+            }
+            prev = &peer;
+         }
+      }
+
+      const controller& cc = self()->chain_plug->chain();
+      auto is_peer_key_valid = [&](const gossip_bp_peers_message::bp_peer& peer) -> bool {
+         try {
+            std::optional<peer_info_t> peer_info = cc.get_peer_info(peer.producer_name);
+            if (peer_info && peer_info->key) {
+               public_key_type pk(peer.sig, peer.digest());
```

Comment:
I think we want to immediately kick out SIG_WA. We did make the contract disallow PUB_WA.

Reply:
https://github.com/AntelopeIO/spring/pull/1265/commits/f8166430a538187d3fd8085b0e7730b0efc6679f